File: pareto_ipca/scripts/proxy_config.R
# proxy_config.R — sourceado pelos scripts que fazem HTTP (seed/reconstruct).
# Edite os campos abaixo na máquina institucional. Sem este arquivo (ou com
# PROXY_HOST = "" abaixo), os scripts rodam SEM proxy — útil em casa.
#
# Como funciona:
#  - Sys.setenv() configura http_proxy/https_proxy pro curl (libcurl) inteiro
#  - httr::set_config(use_proxy(...)) força o httr a usar isso explicitamente
#  - Sem auth: deixe PROXY_USER / PROXY_PASS vazios
#  - Com auth básica: preencha USER/PASS (vai pra URL como user:pass@host:port)
#  - Auth NTLM/Kerberos integrada ao Windows: deixe vazio (sistema resolve)
#  - ssl_verifypeer = 0 só se proxy faz MITM e o cert corporativo não está no
#    bundle do curl. Comece com 1 (verifica); se der "SSL certificate problem",
#    aí muda pra 0.

PROXY_HOST <- ""          # ex.: "proxy.minhainstituicao.br"
PROXY_PORT <- ""          # ex.: "8080"
PROXY_USER <- ""          # vazio se não tiver auth básica
PROXY_PASS <- ""          # vazio se não tiver auth básica
SSL_VERIFY <- TRUE        # mude pra FALSE só se cert corporativo bloquear

# ---------------------------------------------------------------------------
# Não mexa daqui pra baixo.

if (nzchar(PROXY_HOST) && nzchar(PROXY_PORT)) {
  proxy_url <- if (nzchar(PROXY_USER)) {
    sprintf("http://%s:%s@%s:%s",
            utils::URLencode(PROXY_USER, reserved = TRUE),
            utils::URLencode(PROXY_PASS, reserved = TRUE),
            PROXY_HOST, PROXY_PORT)
  } else {
    sprintf("http://%s:%s", PROXY_HOST, PROXY_PORT)
  }
  Sys.setenv(http_proxy = proxy_url, https_proxy = proxy_url)
  if (requireNamespace("httr", quietly = TRUE)) {
    httr::set_config(httr::use_proxy(
      url      = PROXY_HOST,
      port     = as.integer(PROXY_PORT),
      username = if (nzchar(PROXY_USER)) PROXY_USER else NULL,
      password = if (nzchar(PROXY_PASS)) PROXY_PASS else NULL
    ), override = FALSE)
  }
  cat(sprintf("[PROXY] usando %s:%s%s\n",
              PROXY_HOST, PROXY_PORT,
              if (nzchar(PROXY_USER)) sprintf(" (user=%s)", PROXY_USER) else ""))
}

if (!isTRUE(SSL_VERIFY) && requireNamespace("httr", quietly = TRUE)) {
  httr::set_config(httr::config(ssl_verifypeer = 0L, ssl_verifyhost = 0L),
                   override = FALSE)
  cat("[PROXY] SSL verify DESABILITADO (cert corporativo). Use só em rede confiável.\n")
}

File: pareto_cpius/scripts/proxy_config.py
# proxy_config.py — sourceado pelos scripts Python que fazem HTTP.
# Analogo do proxy_config.R (mesmo padrao dos pipelines R).
#
# Edite os campos abaixo na maquina institucional. Sem este arquivo (ou com
# PROXY_HOST = "" abaixo), os scripts rodam SEM proxy — util em casa.
#
# ATENCAO: nao commitar credenciais reais. Este arquivo eh template — mantenha
# os defaults vazios no git. Preencher SO na copia local da maquina corp.
#
# Caso comum (mesmo proxy pra HTTP e HTTPS, sem auth):
#   preenche PROXY_HOST + PROXY_PORT e pronto.
#
# Casos especiais:
#   - Hosts/portas DIFERENTES pra HTTP e HTTPS: preencha HTTPS_PROXY_HOST/PORT.
#     Senao deixa vazio que herda de PROXY_HOST/PORT.
#   - Canal ate o proxy e TLS (raro): mude HTTPS_PROXY_SCHEME pra "https".
#   - Auth basica (user/senha no proxy): preenche PROXY_USER/PASS.
#   - Auth NTLM/Kerberos integrada ao Windows: deixa USER/PASS vazios; sistema
#     resolve via sspi (mas Python nao tem suporte built-in — pode precisar de
#     `cntlm` local como bridge).
#   - Proxy faz MITM e cert corporativo bloqueia: mude SSL_VERIFY pra False.
#     Comeca com True; so desativa se der erro "SSL certificate problem".

# --- HTTP proxy (mandatorio se quiser usar proxy) ---
PROXY_HOST = ""            # ex.: "proxy.minhainstituicao.br"
PROXY_PORT = ""            # ex.: "8080"

# --- HTTPS proxy (opcional: so preenche se for DIFERENTE do HTTP) ---
HTTPS_PROXY_HOST   = ""    # vazio = usa o mesmo PROXY_HOST acima
HTTPS_PROXY_PORT   = ""    # vazio = usa o mesmo PROXY_PORT acima
HTTPS_PROXY_SCHEME = "http"  # quase sempre "http"; so "https" se canal ate proxy for TLS

# --- Autenticacao (opcional) ---
PROXY_USER = ""            # vazio se nao tiver auth basica
PROXY_PASS = ""

# --- Verificacao SSL (deixe True; mude pra False so se proxy fizer MITM) ---
SSL_VERIFY = True

# ---------------------------------------------------------------------------
# Nao mexer daqui pra baixo.
#
# Como o exec() do carregador passa {"os": os} como globals, `os` esta
# disponivel aqui. Setamos env vars maiusculas (HTTPS_PROXY / HTTP_PROXY) —
# urllib/requests respeitam por default.

from urllib.parse import quote as _urlquote  # noqa: E402


def _mk_proxy_url(scheme, host, port, user, pw):
    cred = ""
    if user:
        cred = f"{_urlquote(user, safe='')}:{_urlquote(pw, safe='')}@"
    return f"{scheme}://{cred}{host}:{port}"


if PROXY_HOST and PROXY_PORT:
    _http = _mk_proxy_url("http", PROXY_HOST, PROXY_PORT, PROXY_USER, PROXY_PASS)
    _https_host = HTTPS_PROXY_HOST or PROXY_HOST
    _https_port = HTTPS_PROXY_PORT or PROXY_PORT
    _https = _mk_proxy_url(HTTPS_PROXY_SCHEME, _https_host, _https_port,
                            PROXY_USER, PROXY_PASS)
    # Uppercase eh o que urllib/requests picam por default; lowercase por
    # compat com codigo que usa http_proxy/https_proxy (WSL/curl).
    os.environ["HTTP_PROXY"]  = _http
    os.environ["HTTPS_PROXY"] = _https
    os.environ["http_proxy"]  = _http
    os.environ["https_proxy"] = _https
    print(f"[PROXY] http  {PROXY_HOST}:{PROXY_PORT}")
    _cred_msg = f" (user={PROXY_USER})" if PROXY_USER else ""
    print(f"[PROXY] https {HTTPS_PROXY_SCHEME}://{_https_host}:{_https_port}{_cred_msg}")

if not SSL_VERIFY:
    # Best-effort: seta PYTHONHTTPSVERIFY e desabilita warnings do urllib3.
    # Codigo cliente que usar requests precisa passar verify=False; urllib
    # respeitas PYTHONHTTPSVERIFY=0 na maioria dos casos.
    os.environ["PYTHONHTTPSVERIFY"] = "0"
    try:
        import ssl as _ssl
        _ssl._create_default_https_context = _ssl._create_unverified_context
    except Exception:
        pass
    print("[PROXY] SSL verify DESABILITADO (cert corporativo). Use so em rede confiavel.")
